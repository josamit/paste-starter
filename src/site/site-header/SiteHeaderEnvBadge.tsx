import React from "react";
import { Badge } from "@twilio-paste/core";
import { BadgeVariants } from "@twilio-paste/badge/dist/types";
import { useWindowObject } from "@hooks/useWindowObject";

const determineBadgeVariant = (environment: string) => {
  let variant: BadgeVariants = "error";

  if (environment === "stage") {
    variant = "warning";
  }

  if (environment === "dev") {
    variant = "success";
  }

  return variant;
};

const SiteHeaderEnvBadge: React.FC = () => {
  const { environment } = useWindowObject();

  const variant = determineBadgeVariant(environment);

  return (
    <Badge as="span" variant={variant}>
      {environment}
    </Badge>
  );
};

export default SiteHeaderEnvBadge;
