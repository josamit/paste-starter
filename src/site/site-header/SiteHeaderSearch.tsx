import React, { useState } from "react";
import { Box } from "@twilio-paste/box";
import { Input } from "@twilio-paste/input";
import { SearchIcon } from "@twilio-paste/icons/esm/SearchIcon";
import { Button } from "@twilio-paste/button";
import { useWindowSize } from "@hooks/useWindowSize";
import { useLocation } from "react-router-dom";
import { useQueryManager } from "@context/QueryManagerContext";
import { useToasts } from "@context/ToastsProvider";

const SiteHeaderSearch: React.VFC = () => {
  const toaster = useToasts();
  const { pathname } = useLocation();

  const { breakpointIndex } = useWindowSize();
  const { queryParams, setQueryParams } = useQueryManager();

  const [value, setValue] = useState<string>(queryParams.accountSid || "");
  const [hasError, setHasError] = useState(false);

  React.useEffect(() => {
    if (queryParams.accountSid && value !== queryParams.accountSid.trim()) {
      setValue(queryParams.accountSid);
    }
  }, [queryParams.accountSid]);

  const isMobileNav = breakpointIndex !== undefined && breakpointIndex < 2;

  return (
    <Box minWidth={["size30", "size30", "size30", "size40"]}>
      <form
        onSubmit={(e) => {
          e.preventDefault();
          setHasError(false);
          // @ts-ignore
          const inputText = e.target?.search_input?.value || value;

          if (!inputText || inputText.trim().length === 0) {
            setHasError(true);
            return;
          }

          setQueryParams(pathname, { accountSid: inputText });
        }}
      >
        <Input
          type="text"
          id="search_input"
          name="search_input"
          data-testid="site_header_search_input"
          placeholder="Search by account"
          value={value}
          onChange={(event) => {
            setValue(event.target.value);
          }}
          variant={isMobileNav ? "default" : "inverse"}
          insertAfter={
            <Box cursor="default">
              <Button
                type="submit"
                variant="reset"
                data-testid="site_header_search_btn"
              >
                <SearchIcon decorative />
              </Button>
            </Box>
          }
          hasError={hasError}
        />
      </form>
    </Box>
  );
};

export { SiteHeaderSearch };
