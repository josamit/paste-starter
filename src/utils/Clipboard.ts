import { UseToasterReturnedProps } from "@twilio-paste/core";

const handleErr = (
  toaster: UseToasterReturnedProps,
  itemName: string,
  err: unknown
) => {
  toaster.push({
    message: `Oops, unable to copy ${itemName} to clipboard. ${err}`,
    variant: "error",
  });
};

const handleSuccess = (toaster: UseToasterReturnedProps, message: string) => {
  toaster.push({
    message: `${message}`,
    variant: "success",
  });
};

const handleWarning = (toaster: UseToasterReturnedProps, message: string) => {
  toaster.push({
    message: `${message}`,
    variant: "warning",
  });
};

export function copyToClipboardSync(
  textToCopy: string,
  itemName: string,
  toaster: UseToasterReturnedProps
) {
  const textArea = document.createElement("textarea");
  textArea.value = textToCopy;

  document.body.appendChild(textArea);
  textArea.focus();
  textArea.select();

  try {
    const copyResult = document.execCommand("copy");
    const message = copyResult
      ? `${itemName} was successfully copied to your clipboard.`
      : `${itemName} could not be copied to your clipboard. Please try again.`;
    if (copyResult) {
      handleSuccess(toaster, message);
    } else {
      handleWarning(toaster, message);
    }
  } catch (err) {
    handleErr(toaster, itemName, err);
  }

  document.body.removeChild(textArea);
}

export function copyToClipboardAsync(
  textToCopy: string,
  itemName: string,
  toaster: UseToasterReturnedProps
) {
  if (!navigator.clipboard) {
    copyToClipboardSync(textToCopy, itemName, toaster);
    return;
  }

  navigator.clipboard
    .writeText(textToCopy)
    .then(() =>
      handleSuccess(
        toaster,
        `${itemName} was successfully copied to your clipboard.`
      )
    )
    .catch((err) => handleErr(toaster, itemName, err));
}
