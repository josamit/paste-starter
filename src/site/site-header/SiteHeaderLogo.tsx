import * as React from "react";
import { Text } from "@twilio-paste/text";
import { Box } from "@twilio-paste/box";
import {
  MediaBody,
  MediaFigure,
  MediaObject,
} from "@twilio-paste/media-object";
import { styled } from "@twilio-paste/styling-library";
import { Link } from "react-router-dom";
import { LogoTwilioIcon } from "@twilio-paste/icons/esm/LogoTwilioIcon";
import { Stack } from "@twilio-paste/stack";
import SiteHeaderEnvBadge from "./SiteHeaderEnvBadge";

const StyledLink = styled(Link)`
  text-decoration: none;
  color: inherit;

  &:visited {
    color: inherit;
  }
`;

interface SiteHeaderLogoProps {
  title?: string;
  subtitle?: string;
}

// Note: 'subtitle' isn't passed for the mobile view, so we use that fact
// to render different sizes and spacing in mobile
const SiteHeaderLogo: React.VFC<SiteHeaderLogoProps> = ({
  title,
  subtitle,
}) => {
  return (
    <Box
      display="flex"
      alignItems="center"
      paddingRight="space70"
      paddingLeft="space70"
      borderRightColor="colorBorderInverseWeaker"
      borderRightStyle="solid"
      borderRightWidth={["borderWidth0", "borderWidth0", "borderWidth10"]}
      minWidth={subtitle ? "sizeSidebar" : "size0"}
    >
      <MediaObject verticalAlign="center">
        <MediaFigure spacing="space40">
          <LogoTwilioIcon
            decorative
            color="colorTextInverse"
            size="sizeIcon60"
          />
        </MediaFigure>
        <MediaBody>
          <Text
            as="div"
            fontSize="fontSize40"
            lineHeight="lineHeight40"
            color="colorTextInverse"
          >
            <StyledLink to="/">{title}</StyledLink>
          </Text>
          {subtitle ? (
            <Stack orientation="horizontal" spacing="space20">
              <Text
                as="span"
                fontSize="fontSize20"
                fontWeight="fontWeightNormal"
                lineHeight="lineHeight20"
                color="colorTextInverse"
              >
                {subtitle}
              </Text>
            </Stack>
          ) : null}
        </MediaBody>
      </MediaObject>
    </Box>
  );
};

export { SiteHeaderLogo };
