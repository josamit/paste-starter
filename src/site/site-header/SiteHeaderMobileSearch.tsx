import * as React from "react";
import { Box } from "@twilio-paste/box";
import { Heading } from "@twilio-paste/heading";
import { pasteBaseStyles } from "@twilio-paste/theme";
import { Button } from "@twilio-paste/button";
import { CloseIcon } from "@twilio-paste/icons/esm/CloseIcon";
import { css, styled } from "@twilio-paste/styling-library";
import {
  ModalDialogPrimitiveContent,
  ModalDialogPrimitiveOverlay,
} from "@twilio-paste/modal-dialog-primitive";
import { SiteHeaderSearch } from "./SiteHeaderSearch";

const StyledModalDialogOverlay = styled(ModalDialogPrimitiveOverlay)(
  // @ts-ignore
  css({
    position: "fixed",
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    padding: "space30",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "colorBackgroundOverlay",
    zIndex: "zIndex80",
    overflow: "scroll",
  }),
  // import Paste Theme Based Styles due to portal positioning.
  // reach portal is a sibling to the main app, so you are now
  // no longer a child of the theme provider. We need to re-set
  // some of the base styles that we rely on inheriting from
  // such as font-family and line-height so that compositions
  // of paste components in the modal are styled correctly
  pasteBaseStyles
);

// @ts-ignore
const StyledModalDialogContent = styled(ModalDialogPrimitiveContent)(
  // @ts-ignore
  css({
    position: "absolute",
    top: 10,
    width: "calc(100% - 20px)",
    maxWidth: "500px",
    height: "200px",
    paddingY: "space70",
    paddingX: "space60",
    backgroundColor: "colorBackgroundBody",
    borderRadius: "borderRadius20",
  })
);

interface MobileSearchProps {
  isOpen: boolean;
  onClose: () => void;
}

const MobileSearch: React.FC<MobileSearchProps> = ({ isOpen, onClose }) => {
  if (!isOpen) {
    return null;
  }

  return (
    <StyledModalDialogOverlay onDismiss={onClose} allowPinchZoom>
      <Box
        height="300px"
        width="100vw"
        overflow="scroll"
        display="flex"
        justifyContent="center"
      >
        <StyledModalDialogContent aria-label="Website navigation">
          <Box position="absolute" top="24px" right="24px" zIndex="zIndex10">
            <Button variant="reset" size="reset" onClick={onClose}>
              <CloseIcon
                decorative={false}
                color="colorTextWeak"
                size="sizeIcon60"
                title="Close Search"
              />
            </Button>
          </Box>
          <Box
            display="flex"
            flexDirection="column"
            alignItems="center"
            justifyContent="center"
            position="relative"
          >
            <Heading as="label" variant="heading30">
              Search
            </Heading>
            <SiteHeaderSearch />
          </Box>
        </StyledModalDialogContent>
      </Box>
    </StyledModalDialogOverlay>
  );
};
export { MobileSearch };
