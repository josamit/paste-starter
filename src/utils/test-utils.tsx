import * as React from "react";
import { render as rtlRender } from "@testing-library/react";
import { Theme } from "@twilio-paste/core/theme";
import { MockQueryManagerContext } from "@context/QueryManagerContext";
import { WindowSizeContext } from "@context/WindowSizeProvider";
import { MemoryRouter, Route, Switch } from "react-router-dom";
import { DarkModeContext } from "@context/DarkModeContext";
import ToastsProvider from "@context/ToastsProvider";
import { createGetForm } from "@src/utils/form";
import { Box } from "@twilio-paste/core/dist/box";

export function renderWithFormUtils(...args: Parameters<typeof rtlRender>) {
  const [children, ...restArgs] = args;

  const results = rtlRender(<Box>{children}</Box>, ...restArgs);
  const getForm = createGetForm(results);

  return {
    ...results,
    getForm,
  };
}

const renderComponent = (
  ui: React.ReactElement,
  {
    initialEntry = "/route/1",
    initialRoute = "/route/:appSid",
    queryParams = {
      accountSid: "ACxxx",
      activeTab: "a2pBrandsReconciliation",
    },
    container = undefined,
    mockToggleMode = jest.fn(),
    setQueryParams = jest.fn(),
    ...renderOptions
  } = {}
) => {
  const Wrapper: React.FC = ({ children }) => (
    <MockQueryManagerContext
      value={{
        queryParams,
        setQueryParams,
      }}
    >
      <WindowSizeContext.Provider
        value={{ width: 1400, height: 800, breakpointIndex: 450 }}
      >
        <Theme.Provider theme="default" disableAnimations>
          <DarkModeContext.Provider
            value={{
              theme: "default",
              toggleMode: mockToggleMode,
              componentMounted: true,
            }}
          >
            <ToastsProvider>
              <MemoryRouter initialEntries={[initialEntry]}>
                <Switch>
                  <Route path={initialRoute}>{children}</Route>
                </Switch>
              </MemoryRouter>
            </ToastsProvider>
          </DarkModeContext.Provider>
        </Theme.Provider>
      </WindowSizeContext.Provider>
    </MockQueryManagerContext>
  );

  return renderWithFormUtils(ui, {
    wrapper: Wrapper,
    container,
    ...renderOptions,
  });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { renderComponent };
