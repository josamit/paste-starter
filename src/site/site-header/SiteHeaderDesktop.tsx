import * as React from "react";
import { Box } from "@twilio-paste/box";
import { Stack } from "@twilio-paste/stack";
import { SiteHeaderSearch } from "./SiteHeaderSearch";
import { SiteHeaderLogo } from "./SiteHeaderLogo";
import { DarkModeToggle } from "./DarkModeToggle";

const SiteHeaderDesktop: React.VFC = () => {
  return (
    <Box
      as="header"
      display={["none", "none", "flex"]}
      alignItems="stretch"
      backgroundColor="colorBackgroundBodyInverse"
      borderBottomColor="colorBorderInverseWeaker"
      borderBottomWidth="borderWidth10"
      borderBottomStyle="solid"
      flex="1 0 auto"
    >
      <SiteHeaderLogo title="Paste Starter" subtitle="Admin template" />
      <Box
        display="flex"
        justifyContent="space-between"
        alignItems="center"
        flexDirection="row"
        paddingTop="space60"
        paddingRight="space70"
        paddingBottom="space60"
        paddingLeft="space70"
        width="100%"
      >
        <DarkModeToggle />
        <Box marginTop="space10">
          <Stack orientation="horizontal" spacing="space60">
            <SiteHeaderSearch />
          </Stack>
        </Box>
      </Box>
    </Box>
  );
};

export { SiteHeaderDesktop };
