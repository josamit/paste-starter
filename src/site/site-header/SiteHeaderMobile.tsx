import * as React from "react";
import { Box } from "@twilio-paste/box";
import { Button } from "@twilio-paste/button";
import { SearchIcon } from "@twilio-paste/icons/esm/SearchIcon";
import { HamburgerToggle } from "./HamburgerToggle";
import { SiteHeaderLogo } from "./SiteHeaderLogo";
import { MobileNavigation } from "./SiteHeaderMobileNavigation";
import { MobileSearch } from "./SiteHeaderMobileSearch";

const SiteHeaderMobile: React.VFC = () => {
  const [navigationOpen, setNavigationOpen] = React.useState(false);
  const [searchOpen, setSearchOpen] = React.useState(false);

  return (
    <Box
      as="header"
      display={["flex", "flex", "none"]}
      alignItems="center"
      justifyContent="space-between"
      backgroundColor="colorBackgroundBodyInverse"
      position="sticky"
      top="0px"
      zIndex="zIndex80"
    >
      <Box margin="space60">
        <Button
          variant="inverse"
          size="icon_small"
          onClick={() => setSearchOpen(!searchOpen)}
        >
          <Box paddingX="space20">
            <SearchIcon
              size="sizeIcon50"
              decorative={false}
              title="Search by Account SID (ACxxx)"
            />
          </Box>
        </Button>
        <MobileSearch
          isOpen={searchOpen}
          onClose={() => setSearchOpen(false)}
        />
      </Box>
      <SiteHeaderLogo title="Paste Starter" />
      <Box margin="space60">
        <Button
          data-testid="open_main_navigation"
          aria-label="Open main navigation"
          variant="inverse"
          size="icon_small"
          onClick={() => setNavigationOpen(!navigationOpen)}
        >
          <HamburgerToggle
            aria-hidden
            toggled={navigationOpen}
            color={
              navigationOpen
                ? "colorBackgroundBodyInverse"
                : "colorBackgroundBody"
            }
          />
        </Button>
        <MobileNavigation
          isOpen={navigationOpen}
          onClose={() => setNavigationOpen(false)}
        />
      </Box>
    </Box>
  );
};

export { SiteHeaderMobile };
