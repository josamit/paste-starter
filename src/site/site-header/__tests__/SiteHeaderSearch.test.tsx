import * as React from "react";
import { renderComponent } from "@src/utils/test-utils";
import { SiteHeaderSearch } from "@site/site-header/SiteHeaderSearch";
import { act, fireEvent, waitFor } from "@testing-library/react";

describe("<SiteHeaderSearch />", () => {
  it("should render correctly", async () => {
    const { container } = renderComponent(<SiteHeaderSearch />);

    expect(container).toMatchSnapshot();
  });

  it("should be able to set the account sid correctly", async () => {
    const mockSetQueryParams = jest.fn();
    const { container, getByTestId } = renderComponent(<SiteHeaderSearch />, {
      setQueryParams: mockSetQueryParams,
    });

    await act(() => {
      fireEvent.input(getByTestId("site_header_search_input"), {
        target: { value: "ACyyy" },
      });
    });

    await act(() => {
      fireEvent.click(getByTestId("site_header_search_btn"));
    });

    await waitFor(() => {
      expect(mockSetQueryParams).toHaveBeenNthCalledWith(1, "/route/1", {
        accountSid: "ACyyy",
      });
    });
  });
});
