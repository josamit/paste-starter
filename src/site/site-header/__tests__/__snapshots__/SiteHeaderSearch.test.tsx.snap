// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SiteHeaderSearch /> should render correctly 1`] = `
<div>
  <div
    class="css-ql39e"
  >
    <div
      class="css-1oex4ez"
      data-paste-element="BOX"
    >
      <form>
        <div
          class="css-1jbznaz"
          color="colorTextInverse"
          cursor="text"
          data-paste-element="INPUT"
          display="flex"
          width="100%"
        >
          <input
            aria-invalid="false"
            class="css-dw9535"
            color="inherit"
            cursor="auto"
            data-paste-element="INPUT_ELEMENT"
            data-testid="site_header_search_input"
            display="block"
            font-family="inherit"
            font-size="fontSize30"
            font-weight="fontWeightMedium"
            id="search_input"
            name="search_input"
            placeholder="Search by account"
            type="text"
            value="ACxxx"
            width="100%"
          />
          <div
            class="css-14mvu2x"
            data-paste-element="INPUT_SUFFIX"
            display="flex"
          >
            <div
              class="css-1os76sl"
              cursor="default"
              data-paste-element="BOX"
            >
              <button
                aria-busy="false"
                class="css-tge1d4"
                color="inherit"
                cursor="pointer"
                data-paste-element="BUTTON"
                data-testid="site_header_search_btn"
                display="inline-block"
                font-family="fontFamilyText"
                font-size="inherit"
                font-weight="inherit"
                text-decoration="none"
                type="submit"
                width="auto"
              >
                <span
                  aria-hidden="false"
                  class="css-beroof"
                  data-paste-element="BOX"
                  display="flex"
                  opacity="1"
                  text-decoration="inherit"
                >
                  <span
                    class="css-zeqpzn"
                    color="currentColor"
                    data-paste-element="ICON"
                    display="block"
                  >
                    <svg
                      aria-hidden="true"
                      height="100%"
                      role="img"
                      viewBox="0 0 20 20"
                      width="100%"
                    >
                      <path
                        d="M5.43 5.43a4.882 4.882 0 017.383 6.347l2.973 2.973a.732.732 0 01-1.036 1.036l-2.973-2.973A4.883 4.883 0 015.43 5.43zm1.035 1.035a3.417 3.417 0 104.833 4.833 3.417 3.417 0 00-4.833-4.833z"
                        fill="currentColor"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
